eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -0,0 +1,886 @@
+  where
+    toCompactValue :: CompactForm Coin -> CompactForm (Core.Value era)
+    toCompactValue ada =
+      fromMaybe (error "Failed to compact a `Coin` as `CompactForm (Core.Value era)`")
+        . toCompact
+        . inject
+        $ fromCompact ada
+
+viewTxOut ::
+  forall era.
+  Era era =>
+  TxOut era ->
+  (Addr (Crypto era), Core.Value era, StrictMaybe (DataHash (Crypto era)))
+viewTxOut (TxOutCompact bs c) = (addr, val, SNothing)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+viewTxOut (TxOutCompactDH bs c dh) = (addr, val, SJust dh)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+viewTxOut (TxOutCompactDatum bs c datum) = (addr, val, SJust dh)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+    dh = hashData datum
+viewTxOut (TxOut_AddrHash28_AdaOnly stakeRef a b c d adaVal) =
+  (addr, inject (fromCompact adaVal), SNothing)
+  where
+    addr = decodeAddress28 stakeRef a b c d
+viewTxOut (TxOut_AddrHash28_AdaOnly_DataHash32 stakeRef a b c d adaVal e f g h) =
+  (addr, inject (fromCompact adaVal), SJust (decodeDataHash32 e f g h))
+  where
+    addr = decodeAddress28 stakeRef a b c d
+
+instance
+  ( Era era,
+    Show (Core.Value era),
+    Show (CompactForm (Core.Value era))
+  ) =>
+  Show (TxOut era)
+  where
+  show = show . viewTxOut
+
+deriving via InspectHeapNamed "TxOut" (TxOut era) instance NoThunks (TxOut era)
+
+pattern TxOut ::
+  forall era.
+  ( Era era,
+    Compactible (Core.Value era),
+    Val (Core.Value era),
+    HasCallStack
+  ) =>
+  Addr (Crypto era) ->
+  Core.Value era ->
+  StrictMaybe (DataHash (Crypto era)) ->
+  TxOut era
+pattern TxOut addr vl dh <-
+  (viewTxOut -> (addr, vl, dh))
+  where
+    TxOut (Addr network paymentCred stakeRef) vl SNothing
+      | StakeRefBase stakeCred <- stakeRef,
+        Just adaCompact <- getAdaOnly (Proxy @era) vl,
+        Just (Refl, a, b, c, d) <- encodeAddress28 network paymentCred =
+        TxOut_AddrHash28_AdaOnly stakeCred a b c d adaCompact
+    TxOut (Addr network paymentCred stakeRef) vl (SJust dh)
+      | StakeRefBase stakeCred <- stakeRef,
+        Just adaCompact <- getAdaOnly (Proxy @era) vl,
+        Just (Refl, a, b, c, d) <- encodeAddress28 network paymentCred,
+        Just (Refl, e, f, g, h) <- encodeDataHash32 dh =
+        TxOut_AddrHash28_AdaOnly_DataHash32 stakeCred a b c d adaCompact e f g h
+    TxOut addr vl mdh =
+      let v = fromMaybe (error "Illegal value in txout") $ toCompact vl
+          a = compactAddr addr
+       in case mdh of
+            SNothing -> TxOutCompact a v
+            SJust dh -> TxOutCompactDH a v dh
+
+{-# COMPLETE TxOut #-}
+
+-- ======================================
+
+type ScriptIntegrityHash crypto = SafeHash crypto EraIndependentScriptIntegrity
+
+data TxBodyRaw era = TxBodyRaw
+  { _inputs :: !(Set (TxIn (Crypto era))),
+    _collateral :: !(Set (TxIn (Crypto era))),
+    _collateralAddresses :: !(StrictMaybe (Addr (Crypto era))),
```

Comment:
This should be singular (there's only one) and should have a more helpful name - maybe `collateralReturn`, to indicate its function